modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java

```diff
@@ -0,0 +1,391 @@
+        int latestCatalogVersion = catalogManager.latestCatalogVersion();
+        Optional<CatalogZoneDescriptor> zoneDesciptorOptional = catalogManager.zones(latestCatalogVersion).stream()
+                .filter(catalogZoneDescriptor -> catalogZoneDescriptor.name().equals(zoneName))
+                .findAny();
+
+        if (zoneDesciptorOptional.isEmpty()) {
+            return CompletableFuture.failedFuture(new DistributionZoneNotFoundException(zoneName, null));
+        }
+
+        CatalogZoneDescriptor zoneDescriptor = zoneDesciptorOptional.get();
+
+        Set<NodeWithAttributes> logicalTopology = dzManager.logicalTopology();
+
+        LocalPartitionStatesRequest localPartitionStatesRequest = MSG_FACTORY.localPartitionStatesRequest()
+                .zoneId(zoneDescriptor.id())
+                .catalogVersion(latestCatalogVersion)
+                .build();
+
+        Map<TablePartitionId, Map<String, LocalPartitionState>> result = new ConcurrentHashMap<>();
+        List<CompletableFuture<?>> futures = new ArrayList<>();
+
+        for (NodeWithAttributes node : logicalTopology) {
+            CompletableFuture<NetworkMessage> invokeFuture = messagingService.invoke(
+                    node.nodeName(),
+                    localPartitionStatesRequest,
+                    TimeUnit.SECONDS.toMillis(TIMEOUT)
+            );
+
+            futures.add(invokeFuture.thenAccept(networkMessage -> {
+                assert networkMessage instanceof LocalPartitionStatesResponse : networkMessage;
+
+                var response = (LocalPartitionStatesResponse) networkMessage;
+
+                for (LocalPartitionState state : response.states()) {
+                    result.compute(state.partitionId().asTablePartitionId(), (tablePartitionId, map) -> {
+                        if (map == null) {
+                            return Map.of(node.nodeName(), state);
+                        }
+
+                        map = new HashMap<>(map);
+                        map.put(node.nodeName(), state);
+                        return map;
+                    });
+                }
+            }));
+        }
+
+        return CompletableFuture.allOf(futures.toArray(CompletableFuture[]::new)).handle((unused, throwable) -> normalize(result));
+    }
+
+    /**
+     * Creates new operation future, associated with the request, and writes it into meta-storage.
+     *
+     * @param request Request.
+     * @return Operation future.
+     */
+    private CompletableFuture<Void> processNewRequest(ManualGroupUpdateRequest request) {
+        UUID operationId = request.operationId();
+
+        CompletableFuture<Void> operationFuture = new CompletableFuture<Void>()
+                .whenComplete((v, throwable) -> ongoingOperationsById.remove(operationId))
+                .orTimeout(TIMEOUT, TimeUnit.SECONDS);
+
+        ongoingOperationsById.put(operationId, operationFuture);
```

Comment:
Please leave a comment where these futures complete.

Reply:
It's in the javadoc for `DisasterRecoveryManager#ongoingOperationsById`, should I repeat the comment one more time?

Reply:
My bad. no